### ISSUE TITLE
Redesign dispatcher

### ISSUE THREAD

**Issue by RiscadoA** [GameDevTecnico/cubos#241]
- Its hard to specify dependencies between plugins which don't care about each other - `envSettings`, `fileSettings` before `assetsPlugin`.
- should we tag systems to prevent direct dependencies?
- way to group multiple systems?
- make systems or groups of systems run based on conditional systems, which return a `bool` instead of `void`.

The solution could look like:
```cpp
cubos
        .addSystem(mySystem1, mySystem2, mySystem3)
        .tag("assets")
        .after(myOtherSystem)
        .after(All("settings"));
```

Before we start writing code we should write pseudo code of the cars sample using the solution.

**Comment by RiscadoA** [comment]
So, here is my proposal:

```cpp
void windowPlugin(Cubos& cubos)
{
    cubos
        .addStartupSystem(openWindowSystem)
        .tag("window")
        .after("settings")
        .addSystem(pollEventsSystem)
        .tag("poll")
        .addSystem(swapBuffersSystem)
        .tag("swap-buffers");
} 
```

```cpp
void settingsProviderPlugin(Cubos& cubos)
{
    cubos
        .addStartupSystem(provideSettingsSystem)
        .tag("settings");
}
```

```cpp
// Adds a new condition system with tag 'fixed-step' which runs at a fixed time step.
void fixedStepPlugin(Cubos& cubos)
{
   cubos
       .addConditionSystem(returnTrueWhenShouldStepSystem)
       .tag("fixed-step");
}
```

```cpp
void collisionsPlugin(Cubos& cubos)
{
   cubos
        .addPlugin(fixedStepPlugin)
        .beginGroup()
        .tag("sim-step") // tag all systems within the group
        .addSystem(genCollisionEvents)
        (...)
        .endGroup()
}
```

**Comment by RiscadoA** [comment]
If we follow with this, internally, the dispatcher will need to find an execution order for the systems, and also decide when to commit commands. There are three types of constraints to consider (ordered by their 'strength':
- a system may need to only be run after the commands of another system have been commited: effectivelly a hard synchronization point. When commands are commited, no system can be running.
- a system may need to only be run after another finishes - a pair of systems may be forced to run sequentially, in a specific order.
- two systems may access the same resources in an incompatible way (write and read, write and write), in which case they must be run sequentially.

**Comment by RiscadoA** [comment]
Some systems may need to be run on a specific thread (e.g., event pooling, and stuff which uses the render device)

**Comment by RiscadoA** [comment]
To make this multithreaded, we should create a `ThreadPool` object of some kind which is then used by the dispatcher on `Dispatcher::run()`. Threads in the pool should be identified in some manner.

So, to force a system to run on a specific thread, which should provide a method `.onThread`. Example usage:
```cpp
cubos.addStartupSystem(openWindow).onThread("window");
// ...
cubos.addSystem(drawStuff).onThread("window");
```

This complicates things because startup systems and systems are split into 2 dispatchers, so maybe `Cubos` itself should handle converting thread identifiers from strings to a specific thread in the thread pool.

**Comment by RiscadoA** [comment]
Also: on the case of the window it needs to be run specifically on the main thread, so we should maybe reserve the `"main"` id to identify it (or add `.onMainThread`)

**Comment by rsubtil** [comment]
It looks like we will have more and more settings that each system/stage needs to run under. I don't think the current system of specifying all information in a call chain will scale well into the future, and it may be a source of errors for developers.

Generally speaking I think we need to apply concepts of [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) (directed acyclic graph), to both structure the code and to automatically ensure certain properties of the graph through algorithms (e.g. no cyclic references, dependencies solved, etc...)

I've taken the main requirements of our dispatcher, and added my comments on the proposed solution.

----
### Stages need to be able to specify constraints on their usage, such as conditions for running, should they run multithreaded or not, etc...

Stages are growing in configuration steps to properly determine their execution mode and order. I don't think the current form and proposal of specifying them is viable in the long term; it creates long chains of function calls and can create implicit errors:
```cpp
cubos.addSystem(physics, "Physics")
    .addCondition(physics_fixed_tick)
    .threadedMode(MODE_MULTITHREADED)
    .runsAfter("Physics", "Update")
```
The calls following `addSystem` refer to the currently added system. So, `cubos` needs to keep an internal state of what's currently doing. Unless this state is reset for new systems, these settings can "leak" into another system/stage specification if the developer doesn't specify every bit of configuration:
```cpp
cubos.addSystem(windowPoll, "Poll")
    .threadedMode(MODE_MAIN_THREAD) // Should only run on the main thread
    ...
    .addSystem(updateEnemies, "UpdateEnemies")
    ...	// If user forgets to specify threading mode, this will run in single thread.
```
Furthermore, some systems may share some settings, and it's cumbersome for the user to specify all these over again. The proposed solution of having `beginGroup/endGroup` further creates an internal state that, if not handled by the developer, further leaks into future definitions:
```cpp
cubos.beginGroup()
    .addSystem(updateEnemies, "updateEnemies")
    .tag("update")
    .addPlugin(myCustomPlugin) // This plugin will inherit current configs, e.g. tag "update"
    ...
    .endGroup()
```

**Proposed solution**: Introduce new classes/structs for specifying settings for each stage, making these configurations explicit:

```cpp
class StructSettings
{
public:
    string tag = "";
    ThreadingMode threadingMode = MODE_MULTITHREADED;
    BoolFunction execCondition;
    std::vector<String> dependencies;
    ...
};
...
// Settings need to be initialized before hand
StructSettings physicsSettings = {"Physics", MODE_MULTITHREADED, physics_fixed_tick, {"Update"}};
cubos.addSystem(physicsSystem, physicsSettings);
...
// And settings can be reused for common systems
StructSettings updateSettings;
updateSettings.tag("Update");
cubos.addSystem(updateEnemies, updateSettings)
    .addSystem(updatePlayer, updateSettings);
```
These unique constraints are then formalized in the dispatcher when it determines execution order, and after that step all `StructSettings` can even be eliminated from memory.

----
### Stages can have and provide dependencies that condition how the whole system works. These dependencies need to be "generic", meaning a given stage shouldn't care how exactly that dependency is solved (e.g. the `settings` dependency can be provided either from `envSettings` or `fileSettings`)

I like the proposed tag system, and I think is a good solution for now, but I'd reword it to make sense with the dependency chain (something like `provides`). We only need to define what generic names we want on the engine by default and expose that information to developers so they can properly use them.

----
### The dispatcher will need to automatically handle a dependency chain and figure out in what order to run stages, and parallelize where it can, all without programmer intervention.

This shouldn't be an issue if we model this problem as a graph; we can then apply algorithms to determine execution order, dependency filling, etc...

The dispatcher will receive information on each node about dependencies and providers, and build the graph accordingly. This process needs to result in a unique graph, without ambiguities in it's construction:
![image](https://user-images.githubusercontent.com/6501975/206168627-1d5f25c5-4bf6-472a-b850-eedfb046ee7c.png)
In this example, only a single graph is possible. The only parallelization opportunity is on nodes **C** and **E**. Node **A** is left unused because a newer node **C** was specified and provides the red dependency, replacing **A**.

The developer should be alerted of any problems that exists, in a meaningful way, to understand any error they made and how to fix. Some problems that may arise:

- **Dependency cycle**: If systems depend on each other forming a loop, no system can be fully satisfied. These loops are invalid and should be reported to the user:
`Error: Stages "B; C; D" have a dependency cycle`
![image](https://user-images.githubusercontent.com/6501975/206163642-90b8f799-1029-496b-802f-73b40006d144.png)
- **Dependency satisfaction**: A node that requires a dependency must have that dependency satisfied, otherwise an error occurs:
`Error: Stage "B" has unfulfilled dependency "foo"`
![image](https://user-images.githubusercontent.com/6501975/206164755-c8caa45e-cf28-4225-b881-e394673d0340.png)
- **Parallelization**: Nodes that can run in parallel need to be automatically handled by the dispatcher. This must take into account not only dependencies, but also threading modes. On this example, **A** always needs to run first before any parallelization is possible. When **A** is done, **E** and **C** can run in parallel. Furthermore, **B** and **D** can also run in parallel, but have to first wait for their dependency providers to finish.
![image](https://user-images.githubusercontent.com/6501975/206165534-c4034a49-0e8e-484d-8b94-fac6b942a246.png)

**Comment by RiscadoA** [comment]
First of all, thank you for the taking the time to write all that!
I agree with you fully regarding the DAG representation and the parallelization, but there are still some issues with this design that I think we should discuss further.

## Tags vs Provides

I agree that the `provides` syntax is easier to understand, but it doesn't apply to some use cases.
I think it mainly makes sense for *startup systems*. They only run a single time, so its really important to get the order straight.
For the settings example we've been discussing, where there are systems which *provide* the settings that must run before the systems who use them, it makes sense to use the method names `provides` and `requires`.

The problem is, lets say you want so specify that a system must run before the renderer draws stuff.
How would you specify this dependency with the `provides`/`requires` methods?
The same goes for stuff like the physics system and event handling (although I guess you could say the window polling 'provides' window events).
Maybe we should keep `requires` as an alias for `after`?

One benefit of sticking with tags is that we could use them to group systems and solve the settings duplication problem.

## Tags as Groups

I think that the tags could not only serve to specify dependencies but also for grouping systems in a generic way.
Personally I'm not a fan of the settings object solution because its harder/more confuse to pass parameters via a struct than via a function. Sadly C++ doesn't handle this very well, and the code tends to become confuse.

Instead, I think it should be possible to add 'properties' to both tags and individual systems. When you add a property to a tag, it is automatically added to every system with that tag.

### Syntax

With this architecture we can avoid altogether the `beginGroup()`/`endGroup()` weirdness and settings leaking.
Instead of having super-long chains, chains are only used for each system/tag, to specify its properties. 

```cpp
cubos.tag("fixed-step").runIf(returnsTrueWhenRightTimeStep);

cubos.tag("collision").tagged("fixed-step"); // all 'collision' systems are also 'fixed-step'
cubos.system(detectCollision).tagged("collision");

cubos.tag("physics").tagged("fixed-step").after("collision").before("transform");
cubos.system(doPhysics).tagged("physics");
```

```cpp
cubos.startupSystem(loadEnvSettings).tagged("settings");
cubos.startupSystem(loadFileSettings).tagged("settings");
cubos.startupSystem(windowStuff).after("settings");
```

What this means is that the `tag`/`system` methods don't return a reference to `Cubos`, but to another temporary builder object used to configure the tag/system.
This is good because its consistent with the syntax we're using for `Commands`.

## Startup tags vs normal tags

This brings up the problem of name conflicts between startup system tags and normal system tags.
Here I see two solutions:
- forbid having a startup system and a normal system under the same tag.
- allow startup systems and normal systems to share tags - I see no benefit for this and I think it just complicates everything.

## All properties

So, what properties exactly are there that can be given to systems/tags?
- threading mode - can be optionally be limited to running on the main thread.
- ordering - before/after another system/tag (optionally only after commands have been applied).
- conditions - special systems which must be run before this system/tag and which determine if it will run or not.
- tags - adds a tag to the system(s).

## Good practices

To make this work, all plugins should expose their tags externally so that there are no conflicts.
It would be nice if we could make tags 'type safe' other than just strings to prevent two plugins from using the same tag and thinking its the same. One possible solution to this is to define a type `Tag` that could be used in the following manner:

```cpp
// settings.hpp

const Tag settingsTag;
```

```cpp
// settings.cpp

const Tag settingsTag = Tag("settings"); // name used just for pretty printing purposes
```

```cpp
// env_settings.cpp

#include "settings.hpp"

void envSettingsPlugin(Cubos& cubos)
{
  cubos.startupSystem(loadEnvSettings).tagged(&settingsTag);
}
```

I'm not a fan of this method and personally I think that maybe we should just stick with strings and document all plugins carefully, e.g.:

```cpp
// env_settings.hpp

/// Adds a system tagged with 'settings' which loads settings from environment arguments.
/// etc
void envSettingsPlugin(Cubos& cubos);
```

**Comment by rsubtil** [comment]
As discussed in the meeting, I'll start implementing with the most recent syntax proposed.

**Comment by rsubtil** [comment]
Working on this now, I don't see how the existence of duplicated/shared tags between main/startup dispatchers is an issue. The dispatchers are separated and therefore using the same tag for both will not cause any issues.

**Comment by rsubtil** [comment]
> Yeah, thats true. My concern was that if that happened the user could be confused about how tags work, but if its hard to check I'm fine with us not doing it.

Oh yeah, I forgot about that point. I'll implement it then.

**Comment by RiscadoA** [comment]
I don't think its necessary after reviewing your implementation. If we keep the distinction of `startupTag` vs `tag` ts fine to keep it like it is.
